src/Chainweb/Cut.hs

```diff
@@ -185,42 +201,115 @@ cutWeight :: Getter Cut BlockWeight
 cutWeight = to _cutWeight
 {-# INLINE cutWeight #-}
 
-_cutHeight :: Cut -> BlockHeight
-_cutHeight = sumOf $ cutHeaders . folded . blockHeight
+_cutHeight :: Cut -> CutHeight
+_cutHeight = sumOf $ cutHeaders . folded . blockHeight . to int
 
-cutHeight :: Getter Cut BlockHeight
+cutHeight :: Getter Cut CutHeight
 cutHeight = to _cutHeight
 {-# INLINE cutHeight #-}
 
--- | All adjacent pairs of a cut
---
-_cutAdjPairs :: Cut -> HS.HashSet (AdjPair BlockHeader)
-_cutAdjPairs c = HS.map (fmap (\x -> c ^?! ixg x)) . adjs $ _chainGraph c
+unsafeMkCut :: ChainwebVersion -> HM.HashMap ChainId BlockHeader -> Cut
+unsafeMkCut v hdrs = Cut'
+    { _cutHeaders = hdrs
+    , _cutChainwebVersion = v
+    }
 
-cutAdjPairs :: Getter Cut (HS.HashSet (AdjPair BlockHeader))
-cutAdjPairs = to _cutAdjPairs
-{-# INLINE cutAdjPairs #-}
+-- -------------------------------------------------------------------------- --
+-- Adjacents
+--
+-- Note that in a cut adjacency is directed. The reason is that the headers in a
+-- cut may be of different height and may thus use different graphs.
 
--- | Assumes that the cid exists in the graph.
+-- | Map of adjacent block headers for a chain in a cut. The function considers
+-- the block height for the respective chain in the given cut.
+--
+-- Assumes that the cid exists in the graph.
 --
 cutAdjs
     :: HasChainId cid
     => Cut
     -> cid
     -> HM.HashMap ChainId BlockHeader
-cutAdjs c cid = HM.intersection
-    (_cutHeaders c)
-    (HS.toMap (adjacentChainIds (_chainGraph c) cid))
+cutAdjs c = HM.intersection (_cutHeaders c) . HS.toMap . cutAdjChainIds c
+{-# INLINE cutAdjs #-}
 
-unsafeMkCut :: ChainwebVersion -> HM.HashMap ChainId BlockHeader -> Cut
-unsafeMkCut v hdrs = Cut
-    { _cutHeaders = hdrs
-    , _cutChainwebVersion = v
-    }
+-- | Adjacent chain ids for a chain in a cut. The function considers the block
+-- height for the respective chain in the given cut.
+--
+-- Assumes that the cid exists in the graph.
+--
+cutAdjChainIds :: HasChainId cid => Cut -> cid -> HS.HashSet ChainId
+cutAdjChainIds c cid = c ^?! ixg (_chainId cid) . blockAdjacentChainIds
+{-# INLINE cutAdjChainIds #-}
+
+-- | /Directed/ adjacent pairs of a cut. Note that block headers in a cut can be
+-- of different block height and can thus use different chain graphs.
+--
+_cutAdjPairs :: Cut -> [(BlockHeader, BlockHeader)]
+_cutAdjPairs c =
+    [ (h, x)
```

Comment:
The multiline list comprehension is too weird here. Maybe use `do`?

Reply:
hmm, I find `do` notation for lists much harder to read -- and I imaging that most non-Haskeller programmers would find the same. But in the end it's pretty similar syntax that desugars to the same monadic code. I'll change it to `do` notation.